### PR TITLE
Series pct_change fill_method behavior

### LINKED ISSUE #25006: Series pct_change fill_method behavior
```python
In[2]: import pandas as pd
  ...: import numpy as np
  ...: pd.__version__
Out[2]: u'0.23.4'

In[3]: ts = pd.Series([np.nan, 1., 2., 3., np.nan, 4., np.nan])
In[4]: ts.pct_change(fill_method = None)
Out[4]: 
0    NaN
1    NaN
2    1.0
3    0.5
4    NaN
5    NaN
6    NaN
dtype: float64

In[5]: ts.pct_change(fill_method = 'pad')
Out[5]: 
0         NaN
1         NaN
2    1.000000
3    0.500000
4    0.000000
5    0.333333
6    0.000000
dtype: float64

In[6]: ts.pct_change(fill_method = 'pad').mask(ts.isnull())
Out[6]: 
0         NaN
1         NaN
2    1.000000
3    0.500000
4         NaN
5    0.333333
6         NaN
dtype: float64
```

Hello, 

After recently updating my version, I noticed a change in behavior of pct_change with missing data. This is related to https://github.com/pandas-dev/pandas/issues/19873 . 

First example without fill_method is as expected. The second example is the result now and the third is what it used to be. I think the user should be able to choose if she prefers the second or third behavior. I agree that the second example is correct, as it forward fills as expected, but if the time series is a stock price for example, returns on missing days (holidays) were not 0, which can bias some statistics. 

I would suggest adding a new parameter, like skipna. I could not find any solution with existing parameters, if I missed something please let me know.

Thanks

### PR DESCRIPTION
**Pull request by albertvillanova**
- [ ] closes #25006 
- [ ] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/frame/test_analytics.py [hunk 1]
```diff
@@ -6,6 +6,7 @@
 import warnings
 
 import numpy as np
+from numpy import nan
```
Comment (jreback) [inline pandas/tests/frame/test_analytics.py]: never import nan directly, use np.nan

File: pandas/tests/frame/test_analytics.py [hunk 2]
```diff
@@ -18,6 +19,7 @@
 import pandas.core.algorithms as algorithms
 import pandas.core.nanops as nanops
 import pandas.util.testing as tm
+from pandas.util.testing import (assert_frame_equal)
```
Comment (jreback) [inline pandas/tests/frame/test_analytics.py]: don't use parens like this, you can *already* use ``tm.assert_frame_equal``

**Review by jreback — CHANGES_REQUESTED** [review]
haven't actually looked at the changes, rather some stylistic concerns

File: pandas/core/generic.py [hunk 3]
```diff
@@ -9948,11 +9948,24 @@ def _check_percentile(self, q):
         """
 
     @Appender(_shared_docs['pct_change'] % _shared_doc_kwargs)
-    def pct_change(self, periods=1, fill_method='pad', limit=None, freq=None,
-                   **kwargs):
+    def pct_change(self, periods=1, fill_method=None, limit=None, freq=None,
+                   skipna=False, **kwargs):
+        if skipna and fill_method is not None:
```
Comment (jreback) [inline pandas/core/generic.py]: the fill_method defaults to ``pad`` so not sure how this makes sense, aside from which this is an API change.

also concerned that all other pandas methods have ``skipna=True`` as the default.

Comment (jreback) [reply pandas/core/generic.py]: I think from the OP, having ``skipna=True`` produces 2 while ``skipna=False`` produces 3 yes?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 3 above)
Comment (albertvillanova) [reply pandas/core/generic.py]: @jreback I agree: all other methods have `skipna=True` as default. I just wanted to ask you if I should also set it here. Then, of course, it is an API breaking change, as `fill_method='pad'` will no longer be the default.

BTW, what is the meaning of OP? I do not understand your previous question.

Comment (albertvillanova) [reply pandas/core/generic.py]: @jreback I have a question about setting `skipna=True` as default. Then, in order to use the parameter `fill_method`, the user should pass `skipna=False` explicitly, so that no ValueError is raised. For example:
```python
df.pct_change(fill_method='bfill', limit=2, skipna=False)
```
I do not see this very intuitive. So my question is: should we keep the parameters `fill_method` and `limit` in this method? Or should we better advise to use `fillna` before calling `pct_change`?
For example:
```python
df.fillna(method='bfill', limit=2).pct_change()
```

Comment (albertvillanova) [reply pandas/core/generic.py]: Finally, I have set default `skipna=True` only if `fill_method` and `limit` are None.

**Comment by albertvillanova** [comment]
@jreback all checks have passed.

**Comment by albertvillanova** [comment]
@WillAyd @jschendel could you please have a look?

File: pandas/core/generic.py [hunk 4]
```diff
@@ -9953,11 +9953,24 @@ def _check_percentile(self, q):
...
-                   **kwargs):
+    def pct_change(self, periods=1, fill_method=None, limit=None, freq=None,
+                   skipna=None, **kwargs):
+        if skipna and fill_method is not None:
+            raise ValueError("cannot pass both skipna and fill_method")
+        elif skipna and limit is not None:
+            raise ValueError("cannot pass both skipna and limit")
+        if skipna is None and fill_method is None and limit is None:
+            skipna = True
+        if skipna and self._typ == 'dataframe':
```
Comment (jreback) [inline pandas/core/generic.py]: use isinstance here

File: pandas/core/generic.py [hunk 5]
```diff
@@ -9953,11 +9953,24 @@ def _check_percentile(self, q):
         """
 
     @Appender(_shared_docs['pct_change'] % _shared_doc_kwargs)
-    def pct_change(self, periods=1, fill_method='pad', limit=None, freq=None,
-                   **kwargs):
+    def pct_change(self, periods=1, fill_method=None, limit=None, freq=None,
+                   skipna=None, **kwargs):
+        if skipna and fill_method is not None:
+            raise ValueError("cannot pass both skipna and fill_method")
+        elif skipna and limit is not None:
```
Comment (jreback) [inline pandas/core/generic.py]: do you have tests for each of these conditions?

File: pandas/tests/frame/test_analytics.py [hunk 6]
```diff
@@ -1364,6 +1364,20 @@ def test_pct_change(self):
 
             tm.assert_frame_equal(result, expected)
 
+    @pytest.mark.parametrize("periods, expected_vals", [
+        (1, [[np.nan, np.nan], [np.nan, np.nan], [1., np.nan], [0.5, 1.],
+             [np.nan, 0.5], [0.33333333, np.nan], [np.nan, 0.33333333]]),
+        (2, [[np.nan, np.nan], [np.nan, np.nan], [np.nan, np.nan],
+             [2., np.nan], [np.nan, 2.], [1., np.nan], [np.nan, 1.]])
+    ])
+    def test_pct_change_skipna(self, periods, expected_vals):
```
Comment (jreback) [inline pandas/tests/frame/test_analytics.py]: can you paramaterize these same cases with skipna=True and False

File: pandas/tests/generic/test_generic.py [hunk 7]
```diff
@@ -607,6 +607,22 @@ def test_pct_change(self, periods, fill_method, limit, exp):
...
+        ('pad', None),
+        ('ffill', None),
+        (None, 1)
+    ])
+    def test_pct_change_skipna_raises(self, fill_method, limit):
+        # GH25006
+        if self._typ is DataFrame or self._typ is Series:
+            vals = [np.nan, np.nan, 1, 2, np.nan, 4, 10, np.nan]
+            obj = self._typ(vals)
+            with pytest.raises(ValueError):
```
Comment (jreback) [inline pandas/tests/generic/test_generic.py]: can you check the error message (use match=)

File: pandas/tests/generic/test_generic.py [hunk 8]
```diff
@@ -607,6 +607,22 @@ def test_pct_change(self, periods, fill_method, limit, exp):
...
+    @pytest.mark.parametrize('fill_method, limit', [
+        ('backfill', None),
+        ('bfill', None),
+        ('pad', None),
+        ('ffill', None),
+        (None, 1)
+    ])
+    def test_pct_change_skipna_raises(self, fill_method, limit):
+        # GH25006
+        if self._typ is DataFrame or self._typ is Series:
```
Comment (jreback) [inline pandas/tests/generic/test_generic.py]: this should not be needed any longer, Panels are gone

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py [hunk 9]
```diff
@@ -9953,11 +9953,24 @@ def _check_percentile(self, q):
         """
 
     @Appender(_shared_docs['pct_change'] % _shared_doc_kwargs)
```
Comment (WillAyd) [inline pandas/core/generic.py]: Will certainly want to update the docstring to reflect changes being made here. Probably worth outlining logic in a dedicated Notes section

File: pandas/core/generic.py (hunk 4 above)
Comment (WillAyd) [reply pandas/core/generic.py]: Hmm maybe add this to the frame subclass instead? Somewhat confusing to introspect here in the shared one.

File: pandas/tests/series/test_analytics.py [hunk 10]
```diff
@@ -220,7 +231,7 @@ def test_npdiff(self):
         s = Series(np.arange(5))
 
         r = np.diff(s)
-        assert_series_equal(Series([nan, 0, 0, 0, nan]), r)
+        assert_series_equal(Series([np.nan, 0, 0, 0, np.nan]), r)
```
Comment (WillAyd) [inline pandas/tests/series/test_analytics.py]: OK for now but typically best if you do things like this in separate PRs as its unrelated to the issue you are trying to solve and makes the review process more difficult

File: pandas/tests/frame/test_timeseries.py [hunk 11]
```diff
@@ -143,7 +143,7 @@ def test_pct_change_shift_over_nas(self):
 
         df = DataFrame({'a': s, 'b': s})
 
-        chg = df.pct_change()
+        chg = df.pct_change(fill_method='ffill')
```
Comment (WillAyd) [inline pandas/tests/frame/test_timeseries.py]: Why do we need to add this now? API change?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/generic.py (hunk 5 above)
Comment (albertvillanova) [reply pandas/core/generic.py]: Yes, `test_pct_change_skipna_raises` in tests/generic/test_generic.py.

File: pandas/tests/frame/test_timeseries.py (hunk 11 above)
Comment (albertvillanova) [reply pandas/tests/frame/test_timeseries.py]: Yes, default is skipna=True, as for the rest of pandas methods. See discussion with @jreback in `pandas/core/generic.py`.

File: pandas/core/generic.py (hunk 4 above)
Comment (albertvillanova) [reply pandas/core/generic.py]: For the moment, I have added `isinstance` as required by @jreback. Tell me if you both think I should do otherwise.

File: pandas/core/generic.py [hunk 12]
```diff
@@ -9857,6 +9857,8 @@ def _check_percentile(self, q):
 
         Parameters
         ----------
+        skipna : bool, default True
+            Exclude NA/null values before computing percent change.
```
Comment (jreback) [inline pandas/core/generic.py]: add a versionadded, move this arg to the end

File: pandas/core/generic.py [hunk 13]
```diff
@@ -9904,8 +9906,9 @@ def _check_percentile(self, q):
         2   -0.055556
         dtype: float64
 
-        See the percentage change in a Series where filling NAs with last
-        valid observation forward to next valid.
+        See the computing of percentage change in a Series with NAs. With
+        default skipped NAs, NAs are ignored before the computation and kept
```
Comment (jreback) [inline pandas/core/generic.py]: can you reword this a bit. Its not clear what this is doing. This can be in the Notes section.

File: pandas/core/generic.py [hunk 14]
```diff
@@ -9964,11 +9978,24 @@ def _check_percentile(self, q):
         """
 
     @Appender(_shared_docs['pct_change'] % _shared_doc_kwargs)
-    def pct_change(self, periods=1, fill_method='pad', limit=None, freq=None,
-                   **kwargs):
+    def pct_change(self, skipna=None, periods=1, fill_method=None, limit=None,
+                   freq=None, **kwargs):
+        if skipna and fill_method is not None:
```
Comment (jreback) [inline pandas/core/generic.py]: see my comment about about the ordering of these args

File: pandas/core/generic.py [hunk 15]
```diff
@@ -9964,11 +9978,24 @@ def _check_percentile(self, q):
...
+            raise ValueError("cannot pass both skipna and fill_method")
+        elif skipna and limit is not None:
+            raise ValueError("cannot pass both skipna and limit")
+        if skipna is None and fill_method is None and limit is None:
+            skipna = True
+        if skipna and isinstance(self, pd.DataFrame):
```
Comment (jreback) [inline pandas/core/generic.py]: why is the skipna here? this is strange why you are now handling DataFrame different than before

**Review by jreback — CHANGES_REQUESTED** [review]
this needs a subsection in the whatsnew to show the previous and current behavior.

File: pandas/core/generic.py (hunk 15 above)
Comment (albertvillanova) [reply pandas/core/generic.py]: Yes, DataFrames are handled differently. The problem is that for a DataFrame, the NAs must be dropped independently for each column/row, and this is not possible with `DataFrame.dropna` (thus the need of using `DataFrame.apply` instead). See my question (https://github.com/pandas-dev/pandas/issues/25006#issuecomment-462153438) and the answer by @WillAyd (https://github.com/pandas-dev/pandas/issues/25006#issuecomment-462198723) in the issue #25006.

File: pandas/core/generic.py [hunk 16]
```diff
@@ -9904,22 +9913,42 @@ def _check_percentile(self, q):
...
 
-        See the percentage change in a Series where filling NAs with last
-        valid observation forward to next valid.
+        See how the computing of percentage change is performed in a Series
+        with NAs. With default `skipna=True`, NAs are dropped before the
+        computation and eventually the results are reindexed like the original
+        object, thus keeping the original NAs.
 
-        >>> s = pd.Series([90, 91, None, 85])
+        >>> s = pd.Series([90, 91, None, 85, None, 95])
```
Comment (WillAyd) [inline pandas/core/generic.py]: I realize this was in place before you but let's change `None` to `np.nan` to shy away from inference in the example

File: pandas/core/generic.py [hunk 17]
```diff
@@ -9961,14 +9990,75 @@ def _check_percentile(self, q):
               2016      2015      2014
         GOOG   NaN -0.151997 -0.086016
         APPL   NaN  0.337604  0.012002
+
+        In a DataFrame with NAs, when computing the percentage change with
```
Comment (WillAyd) [inline pandas/core/generic.py]: Can you add a `skipna=False` example to highlight difference?

File: pandas/core/generic.py (hunk 15 above)
Comment (WillAyd) [reply pandas/core/generic.py]: I am generally not a fan of introspection in a shared method - any reason you wouldn't subclass this for DataFrame as required?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.25.0.rst [hunk 18]
```diff
@@ -86,6 +86,7 @@ Other API Changes
 - :class:`DatetimeTZDtype` will now standardize pytz timezones to a common timezone instance (:issue:`24713`)
 - ``Timestamp`` and ``Timedelta`` scalars now implement the :meth:`to_numpy` method as aliases to :meth:`Timestamp.to_datetime64` and :meth:`Timedelta.to_timedelta64`, respectively. (:issue:`24653`)
 - :meth:`Timestamp.strptime` will now rise a ``NotImplementedError`` (:issue:`25016`)
+- Default `skipna=True` for :meth:`Series.pct_change` and :meth:`DataFrame.pct_change` will drop NAs before calculation (:issue:`25006`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.25.0.rst]: this sounds like this is a change, but its actually not, you are just adding the ``skipna`` arg. pls make that more clear. put this in other enhancements.

File: pandas/core/generic.py (hunk 16 above)
Comment (jreback) [reply pandas/core/generic.py]: can you update this

File: pandas/core/generic.py [hunk 19]
```diff
@@ -9961,14 +9990,75 @@ def _check_percentile(self, q):
...
+        5  0.117647  0.000000  0.000000
         """
 
     @Appender(_shared_docs['pct_change'] % _shared_doc_kwargs)
-    def pct_change(self, periods=1, fill_method='pad', limit=None, freq=None,
-                   **kwargs):
-        # TODO: Not sure if above is correct - need someone to confirm.
+    def pct_change(self, periods=1, fill_method=None, limit=None, freq=None,
+                   skipna=None, **kwargs):
+        if fill_method is not None and skipna:
```
Comment (jreback) [inline pandas/core/generic.py]: can you make a bit less add hoc, e.g.

```
if fill_method is None:
      if skipna:
           raise
       if limit is None and skipna is None:
        ......
```

File: pandas/core/generic.py [hunk 20]
```diff
@@ -9961,14 +9990,75 @@ def _check_percentile(self, q):
...
+            raise ValueError("cannot pass both fill_method and skipna")
+        elif limit is not None and skipna:
+            raise ValueError("cannot pass both limit and skipna")
```
Comment (jreback) [inline pandas/core/generic.py]: it would more clear if we an actually default ``skipna=True``, is that probelmatic?

File: pandas/core/generic.py [hunk 21]
```diff
@@ -9961,14 +9990,75 @@ def _check_percentile(self, q):
...
+        if fill_method is None and limit is None and skipna is None:
+            skipna = True
         axis = self._get_axis_number(kwargs.pop('axis', self._stat_axis_name))
-        if fill_method is None:
+        if skipna and isinstance(self, pd.DataFrame):
```
Comment (jreback) [inline pandas/core/generic.py]: use ABCDataFrame

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by TomAugspurger** [comment]
I guess I had a pending review that wasn't submitted:

Is anyone else uncomfortable with making `skipna=True` the default? If I have a timeseries, like

```python
In [9]: s = pd.Series([90, 91, None, 85, None, 95], index=pd.date_range('2000', periods=6))

In [10]: s
Out[10]:
2000-01-01    90.0
2000-01-02    91.0
2000-01-03     NaN
2000-01-04    85.0
2000-01-05     NaN
2000-01-06    95.0
Freq: D, dtype: float64

In [11]: s.pct_change()
Out[11]:
2000-01-01         NaN
2000-01-02    0.011111
2000-01-03         NaN
2000-01-04   -0.065934
2000-01-05         NaN
2000-01-06    0.117647
Freq: D, dtype: float64
```

Ideally, we would read the output as "the percent change between two periods". But, e.g. for the output labels 2000-01-04 the value is -.0.659; That's the percent change between 2000-01-02 and 2000-01-04. I know that's what we do on master, so maybe it's not a big deal... 

To me, the most useful thing with `.pct_chnage` and missing data is for the output to be missing when either the previous or current period is NaN. Can we achieve that with `skipna=False`? Is that documented?

File: pandas/core/generic.py (hunk 17 above)
Comment (albertvillanova) [reply pandas/core/generic.py]: @WillAyd are you sure that example is useful?

```
In [6]: df.pct_change(skipna=False)                                                                                                            
Out[6]: 
          a   b   c
0       NaN NaN NaN
1  0.011111 NaN NaN
2       NaN NaN NaN
3       NaN NaN NaN
4       NaN NaN NaN
5       NaN NaN NaN
```

File: doc/source/whatsnew/v0.25.0.rst (hunk 18 above)
Comment (albertvillanova) [reply doc/source/whatsnew/v0.25.0.rst]: @jreback I am not _just_ adding `skipna` arg; I am setting `skipna=True` as default (before, the default was `fill_method='pad'`). I think this is an API breaking change. Indeed I quoted what you told me:

> the fill_method defaults to pad so not sure how this makes sense, aside from which this is an API change.

File: pandas/core/generic.py (hunk 20 above)
Comment (albertvillanova) [reply pandas/core/generic.py]: Yes, see what I previously commented to you: 
- https://github.com/pandas-dev/pandas/pull/25291#discussion_r257833942

> @jreback I have a question about setting `skipna=True` as default. Then, in order to use the parameter `fill_method`, the user should pass `skipna=False` explicitly, so that no ValueError is raised. For example:
> 
> ```python
> df.pct_change(fill_method='bfill', limit=2, skipna=False)
> ```
> 
> I do not see this very intuitive. So my question is: should we keep the parameters `fill_method` and `limit` in this method? Or should we better advise to use `fillna` before calling `pct_change`?
> For example:
> 
> ```python
> df.fillna(method='bfill', limit=2).pct_change()
> ```


- https://github.com/pandas-dev/pandas/pull/25291#discussion_r258711552

> Finally, I have set default `skipna=True` only if `fill_method` and `limit` are None.

**Comment by albertvillanova** [comment]
@TomAugspurger yes, the behavior you are interested in can be achieved by setting `skipna=False`:

```
In [11]: s = pd.Series([90, 91, None, 85, None, 95, 97], index=pd.date_range('2000', periods=7))                                               

In [12]: s                                                                                                                                     
Out[12]: 
2000-01-01    90.0
2000-01-02    91.0
2000-01-03     NaN
2000-01-04    85.0
2000-01-05     NaN
2000-01-06    95.0
2000-01-07    97.0
Freq: D, dtype: float64

In [13]: s.pct_change(skipna=False)                                                                                                            
Out[13]: 
2000-01-01         NaN
2000-01-02    0.011111
2000-01-03         NaN
2000-01-04         NaN
2000-01-05         NaN
2000-01-06         NaN
2000-01-07    0.021053
Freq: D, dtype: float64
```

I totally agree with you that the behavior with `skipna=False` is much more sensible and intuitive.

File: pandas/core/generic.py (hunk 19 above)
Comment (albertvillanova) [reply pandas/core/generic.py]: @jreback for the moment I leave this part of the code as it is until we agree after @TomAugspurger proposal of setting the default to `skipna=False`. It that is finally the case, all these if conditions would be much simpler.

File: doc/source/whatsnew/v0.25.0.rst (hunk 18 above)
Comment (jreback) [reply doc/source/whatsnew/v0.25.0.rst]: right, maybe also say *adding* the ``skipna`` arg (as its not obvious that it was added in the note)

**Review by jreback — CHANGES_REQUESTED** [review]
i think settingt he default ``skipna=True`` is good. I believe this simplifies the checking a bit as well.

**Comment by WillAyd** [comment]
@albertvillanova can you merge master and address latest comments?

**Comment by jreback** [comment]
can you merge master

**Review by jreback — APPROVED** [review]

File: doc/source/whatsnew/v0.25.0.rst (hunk 18 above)
Comment (jreback) [reply doc/source/whatsnew/v0.25.0.rst]: can you say that thiis is current behavior and the default is NO change.

**Review by jreback — CHANGES_REQUESTED** [review]
@albertvillanova small comment; pls merge master and ping on green.

**Comment by jreback** [comment]
I think only a small comment left, can you merge master and respond.

**Comment by jreback** [comment]
can you merge master here

**Comment by WillAyd** [comment]
Rebased to keep current. Let's see if we can get this one in

**Comment by WillAyd** [comment]
Updated docstring, though on closer look this does change the default behavior against master:

before change:

```python
>>> s = pd.Series([90, 91, np.nan, 85, np.nan, 95])
>>> s.pct_change()
0         NaN
1    0.011111
2    0.000000
3   -0.065934
4    0.000000
5    0.117647
dtype: float64
```

this branch:
```python
>>> s = pd.Series([90, 91, np.nan, 85, np.nan, 95])
>>> s.pct_change()
0         NaN
1    0.011111
2         NaN
3   -0.065934
4         NaN
5    0.117647
dtype: float64
```

So I think need to be careful here

**Comment by TomAugspurger** [comment]
Hmm thanks for checking. This seems like something we can do in a backwards-compatible way (possibly with a deprecation).

**Comment by WillAyd** [comment]
I think this requires some effort to manage the backwards compat piece but not something I personally have time / motivation to dedicate to. Looks stale otherwise so closing, but @albertvillanova if something you'd like to pick back up please ping and we can continue on!
